### PR TITLE
fixing jobs api connection

### PR DESCRIPTION
**Pull request by Njoselson**
This is a continuation of the API fixes as a result of linkedin updating their Voyager API. Specifically this addresses #335   which was also talked about in PR #332 especially the comment [here](https://github.com/tomquirk/linkedin-api/pull/332#issuecomment-1678411347).

This also adds some more testing of the filters than what was there before.

### CODE REVIEW

**Comment by Njoselson** [comment]
Note, there are still a lot of broken API's, but this at least fixes the `search_jobs` endpoints, which were asked for by a lot of people.

**Comment by Njoselson** [comment]
@ignaciovi Can you take a look and see whether this is a better way of formatting the query strings than the long f-strings in the other PR?

File: linkedin_api/linkedin.py [hunk 1]
```diff
@@ -466,53 +466,63 @@ def search_jobs(
...
+        #        distance:List(25),
+        #        company:List(163253),
+        #        salaryBucketV2:List(5),
+        #        timePostedRange:List(r2592000),
+        #        workplaceType:List(1)
+        #    ),
+        #    spellCorrectionEnabled:true
+        #  )"
+
+        query = str(query).replace(" ","").replace("'","").replace("{","(").replace("}",")")
```
Comment (gustavo-momente) [inline linkedin_api/linkedin.py:510]: What happens if the location of the keyword have spaces on them?

**Comment by gustavo-momente** [comment]
It seems that it enters an infinite loop if there are no returned results

**Comment by Njoselson** [comment]
Haha! Good catch! I will fix! 😅

File: linkedin_api/linkedin.py (hunk 1 above)
Comment (Njoselson) [reply linkedin_api/linkedin.py:510]: This just turns "software engineer" into "softwareengineer"... Do you think the linkedin API will care? Maybe it is better to operate on the keywords string first and already transform that into the `software%20engineer`?

Let me fix...

**Comment by Njoselson** [comment]
what do you think now @gustavo-momente ?

File: linkedin_api/linkedin.py (hunk 1 above)
Comment (gustavo-momente) [reply linkedin_api/linkedin.py:510]: I was testing with some query strings with spaces and it was returning no results, but when i used the site I did manage to get results using the same filters. Also, to double check I tried single word queries and was getting results. Probably the behavior depends if linkedin is able to understand the "mistake" and fix it given that we're setting the `spellCorrectionEnabled` flag.

File: linkedin_api/linkedin.py [hunk 2]
```diff
@@ -466,53 +466,63 @@ def search_jobs(
         if limit is None:
             limit = -1
 
-        params = {}
+        query = {"origin":"JOB_SEARCH_PAGE_QUERY_EXPANSION"}
         if keywords:
-            params["keywords"] = keywords
+            query["keywords"] = keywords.replace(" ","%20")
+        if location_name:
+            query["locationFallback"] = location_name.replace(" ","%20")
```
Comment (gustavo-momente) [inline linkedin_api/linkedin.py:473]: Won't we double encode given what what happens on line 525? But I do agree that the solution is something along this lines. Or maybe set something like:

```python
if keywords:
            query["keywords"] = "KEYWORKDS_PLACEHOLDER"
if location_name:
    query["locationFallback"] = "LOCATION_PLACEHOLDER"
```
and then on the replace part (lin 510) you could do something like:
```python
query = str(query).replace(" ","") \
                  .replace("'","") \
                  .replace("KEYWORKDS_PLACEHOLDER", keywords) \
                  .replace("LOCATION_PLACEHOLDER", location_name) \
                  .replace("{","(") \
                  .replace("}",")")
````

File: linkedin_api/linkedin.py [hunk 3]
```diff
@@ -522,10 +532,12 @@ def search_jobs(
...
                     i
                     for i in elements
-                    if i["$type"] == "com.linkedin.voyager.jobs.JobPosting"
+                    if i["$type"] == 'com.linkedin.voyager.dash.jobs.JobPosting'
                 ]
             )
-            # break the loop if we're done searching
+            # break the loop if we're done searching or no results returned
+            if not results:
+                break
```
Comment (gustavo-momente) [inline linkedin_api/linkedin.py:540]: I was testing and found that sometimes this works, others it doesn't. Mostly if you get some results on the first n iterations but on the last you keep getting no results.

I was trying to see what can be done here and what seems to be working is:

````python
new_data = [
   i
   for i in elements
   if i["$type"] == 'com.linkedin.voyager.dash.jobs.JobPosting'
]

if len(new_data) == 0:
   break

results.extend(new_data)
```

That if it finds no new JobPosting in the results it stops iteration.

File: linkedin_api/linkedin.py (hunk 1 above)
Comment (Njoselson) [reply linkedin_api/linkedin.py:510]: if it was the previous implementation then it was probably getting confused by the fact that the spaces were being taken away, so the words were getting mushed together...

File: linkedin_api/linkedin.py (hunk 2 above)
Comment (Njoselson) [reply linkedin_api/linkedin.py:473]: this seems like a nice fix... It makes it a bit more clear what is happening as well..

File: linkedin_api/linkedin.py (hunk 3 above)
Comment (Njoselson) [reply linkedin_api/linkedin.py:540]: ok... That's a nice implementation.

Comment (Njoselson) [reply linkedin_api/linkedin.py]: and of course that makes sense that it wouldn't always work if just testing for "results"

**Comment by tomquirk** [comment]
@Njoselson thanks for this PR! Appreciate the initiative. Do you think it's a net improvement on what we have now? If yes, let's merge and consider improvements/optimizations in a follow-up PR if necessary 🚀

**Comment by Njoselson** [comment]
Hey! Yeah, unfortunately this is also part of the API that is now broken due to updates to the voyager service.

So this is an improvement in that it now passes tests that used to be broken.

**Comment by tomquirk** [comment]
@Njoselson awesome! Thanks again, let's merge 🏁

File: linkedin_api/linkedin.py [hunk 4]
```diff
@@ -466,53 +466,63 @@ def search_jobs(
...
+            query['selectedFilters']['distance'] = f"List({distance})"
+        if remote:
+            query['selectedFilters']['workplaceType'] = f"List({','.join(remote)})"
+
+        query['selectedFilters']['timePostedRange'] = f"List(r{listed_at})"
+        query["spellCorrectionEnabled"] = "true"
+
+        # Query structure:
+        # "(
+        #    origin:JOB_SEARCH_PAGE_QUERY_EXPANSION,
```
Comment (skillIssueSwe) [inline linkedin_api/linkedin.py:497]: Hi, 
I am currently working on a project similar to this one and wanted to ask about how you guys were able to find this origin `JOB_SEARCH_PAGE_QUERY_EXPANSION` as well as the decorationId further down. I know how to find the voyager endpoints but I am pretty stumped how to rediscover endpoints and their needed params if LinkedIn changes their endpoints. 

I did comb through the network feed for awhile and found many origins just not the one here. Any advice is appreciated.

**Comment by egor-ryashin** [comment]
I've just run the test case and it doesn't return anything:
```python
jobs = api.search_jobs(
    keywords="software engineer", 
    location_name="San Francisco",
    companies=["1441","162479"], 
    experience=["1","2","3","4","5","6"],
    job_type=["F","C","P","T","I","V","O"],
    job_title=["9","30006"],
    industries=["24","6"],
    distance=1000,
    remote=["1","2","3"],
    listed_at=1000000,
    limit=10
)
print(jobs)
print(json.dumps(jobs))
```
```
[]
[]
```

**Pull request merged**